exporter/awsemfexporter/cwlog_client.go

```diff
@@ -0,0 +1,237 @@
+						awsErr.Code(),
+						*input.LogGroupName,
+						*input.LogStreamName,
+						awsErr.Message(),
+						awsErr.Error(),
+						awsErr))
+					return token, err
+				}
+				client.logger.Error(fmt.Sprintf("E! cloudwatchlogs: code: %s, message: %s, original error: %#v, %#v", awsErr.Code(), awsErr.Message(), awsErr.OrigErr(), err))
+				return token, err
+			}
+
+		}
+
+		if response != nil {
+			if response.RejectedLogEventsInfo != nil {
+				rejectedLogEventsInfo := response.RejectedLogEventsInfo
+				if rejectedLogEventsInfo.TooOldLogEventEndIndex != nil {
+					client.logger.Warn(fmt.Sprintf("W! %d log events for log group name '%s' are too old", *rejectedLogEventsInfo.TooOldLogEventEndIndex, *input.LogGroupName))
+				}
+				if rejectedLogEventsInfo.TooNewLogEventStartIndex != nil {
+					client.logger.Warn(fmt.Sprintf("W! %d log events for log group name '%s' are too new", *rejectedLogEventsInfo.TooNewLogEventStartIndex, *input.LogGroupName))
+				}
+				if rejectedLogEventsInfo.ExpiredLogEventEndIndex != nil {
+					client.logger.Warn(fmt.Sprintf("W! %d log events for log group name '%s' are expired", *rejectedLogEventsInfo.ExpiredLogEventEndIndex, *input.LogGroupName))
+				}
+			}
+
+			if response.NextSequenceToken != nil {
+				token = response.NextSequenceToken
+				break
+			}
+		}
+	}
+	if err != nil {
+		client.logger.Error("E! All retries failed for PutLogEvents. Drop this request.")
+	}
+	return token, err
+}
+
+//Prepare the readiness for the log group and log stream.
+func (client *cloudWatchLogClient) CreateStream(logGroup, streamName *string) (token string, e error) {
+	//CreateLogStream / CreateLogGroup
+	_, e = callFuncWithRetries(
+		func() (string, error) {
+			_, err := client.svc.CreateLogStream(&cloudwatchlogs.CreateLogStreamInput{
+				LogGroupName:  logGroup,
+				LogStreamName: streamName,
+			})
+			if err != nil {
+				client.logger.Debug(fmt.Sprintf("D! creating stream fail due to : %v \n", err))
+				if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == cloudwatchlogs.ErrCodeResourceNotFoundException {
+					_, err = client.svc.CreateLogGroup(&cloudwatchlogs.CreateLogGroupInput{
+						LogGroupName: logGroup,
+					})
+				}
+
+			}
+			return "", err
+		},
+		cloudwatchlogs.ErrCodeResourceAlreadyExistsException,
+		fmt.Sprintf("E! CreateLogStream / CreateLogGroup with log group name %s stream name %s has errors.", *logGroup, *streamName))
+
+	if e != nil {
+		client.logger.Debug(fmt.Sprintf("D! error != nil, return token: %s with error: %v \n", token, e))
+		return token, e
+	}
+
+	//After a log stream is created the token is always empty.
+	return "", nil
+}
+
+//encapsulate the retry logic in this separate method.
+func callFuncWithRetries(fn func() (string, error), ignoreException string, errorMsg string) (result string, err error) {
+	for i := 0; i <= defaultRetryCount; i++ {
```

Comment:
Put the sleep state here as a local, not in a package variable. Perhaps we don't expect cwlog to be called from multiple threads but there's no reason to expect it won't, the `sleep` global seems to be the only thing preventing thread-safety

Reply:
Though actually, let's just use the same library as queued_retry (assuming this block needs to stay, hoping we can remove retyring from this exporter)

https://github.com/open-telemetry/opentelemetry-collector/blob/1e65674799c8a543e1bc216b8e71de24a374991d/exporter/exporterhelper/queued_retry.go#L22

Reply:
Actually, I think we can remove this retry. The only purpose of this 1-time retry is to make sure it creates log group first and then use 'retry' to create log stream when both of them don't exist. We can do the log stream creation again right after creating log group. In this case, we don't need this retry. 

For other failure cases, we can just throw the error back to the pipeline and let processor handle it.